### DISCUSSION TITLE
Openssl on Windows produces wrong SHA-512 hash

### DISCUSSION THREAD

**Discussion by Tinkerisog** [openssl/openssl#22913] [Q&A] [answered]
Steps to reproduce:
1. go to https://sha512.online/
2. in "INPUT STRING:", type 'password' (without quotes)
3. see the output: b109f3bbbc244eb82441917ed06d618b9008dd09b3befd1b5e07394c706a8bb980b1d7785e5976ec049b46df5f1326af5a2ea6d103fd07c95385ffab0cacbc86
4. go to https://wiki.openssl.org/index.php/Binaries and get openssl-3.2.0.zip from https://www.firedaemon.com/get-openssl
5. unpack zip and run cmd: "echo password| openssl.exe dgst -sha512"
6. see the output: 5139b82e47847baf441c245a136b2fabe64f63be0068bce36c5c736269f119873dc6b56b83e5e59cd736768d62a61e57390b16f18fa72c9f8b9842bc2c3b77e9

**Comment by mspncp** [comment, marked as answer]
It's a classic mistake: The output of the echo command includes the carriage-return/line-feed controls (`\r\n` = `0x0d 0x0a`). Contrary to Linux,  there seems to be no option to turn this off. What you are seeing is the hash of `password\r\n`, as I can show you using the Linux echo command:

```
msp@debian:~$ for pw in password 'password\r\n' ; do echo -ne $pw | hexdump -C ; echo -ne $pw | sha512sum ; echo -ne $pw | openssl dgst -sha512 ; echo; done

00000000  70 61 73 73 77 6f 72 64                           |password|
00000008
b109f3bbbc244eb82441917ed06d618b9008dd09b3befd1b5e07394c706a8bb980b1d7785e5976ec049b46df5f1326af5a2ea6d103fd07c95385ffab0cacbc86  -
SHA2-512(stdin)= b109f3bbbc244eb82441917ed06d618b9008dd09b3befd1b5e07394c706a8bb980b1d7785e5976ec049b46df5f1326af5a2ea6d103fd07c95385ffab0cacbc86

00000000  70 61 73 73 77 6f 72 64  0d 0a                    |password..|
0000000a
5139b82e47847baf441c245a136b2fabe64f63be0068bce36c5c736269f119873dc6b56b83e5e59cd736768d62a61e57390b16f18fa72c9f8b9842bc2c3b77e9  -
SHA2-512(stdin)= 5139b82e47847baf441c245a136b2fabe64f63be0068bce36c5c736269f119873dc6b56b83e5e59cd736768d62a61e57390b16f18fa72c9f8b9842bc2c3b77e9
```


Note also that the Windows echo command ignores all common-sense conventions for argument parsing: as you can see here, the number of trailing spaces before the pipe sign matters, because they are not stripped from the argument.

```
C:\tools\openssl\bin>echo password|openssl dgst -sha512
(stdin)= 5139b82e47847baf441c245a136b2fabe64f63be0068bce36c5c736269f119873dc6b56b83e5e59cd736768d62a61e57390b16f18fa72c9f8b9842bc2c3b77e9

C:\tools\openssl\bin>echo password |openssl dgst -sha512
(stdin)= 89873037d7605c8fdf23ac9a70f4cf5999a13d6256b3fbab44b7941f3cb35e886a14da264fac8f4bc7b192ec48a3f975fc2f2d60bab945d1df64ed29ea87e877

C:\tools\openssl\bin>echo password  |openssl dgst -sha512
(stdin)= 012434295fac93eac3c06b05cd6fb1c3a4bc4f8ecd57dc4017f68d7a2407bd845482967f049dc87dd133ae79560971d2f90201e75e220f8dfc54721f91136495
```

**Reply by Tinkerisogi** [reply]
Thank you so much.
I've figured out a way to get rid of `newline` char on Windows though:
`echo|set /p="password"| openssl.exe dgst -sha512`
Output:
`b109f3bbbc244eb82441917ed06d618b9008dd09b3befd1b5e07394c706a8bb980b1d7785e5976ec049b46df5f1326af5a2ea6d103fd07c95385ffab0cacbc86`

I apologize to OpenSSL community for the noise created.

**Reply by mspncp** [reply]
No need to apologize for asking a serious question. There is no such thing as a stupid question, only a stupid answer.
